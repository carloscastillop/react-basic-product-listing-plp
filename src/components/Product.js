import React from 'react';
import * as Constant from '../Constants';
import '../styles/App.scss';

class Product extends React.Component {
    /**
     *
     * @param option
     * @param id
     * @param price
     */
    updateCart = (option, id, price) => {
        this.props.updateCart(option, id, price);
    }

    /**
     *
     * @param id
     * @param cart
     * @returns {string}
     */
    qtyFilter = (product, cart) => {
        let qty = '';
        if (!cart || !product) {
            return;
        }
        cart.forEach((item) => {
            if (item.hasOwnProperty('id') && product.id === item.id) {
                return qty = item.qty
            }
        });
        return (qty > 0) ? qty : '';
    }

    render() {
        const {product, cart} = this.props;
        const qty = this.qtyFilter(product, cart);
        const disabledState = (qty === '' || qty < 1) ? true : false;

        return (
            <React.Fragment>
                {product &&
                <div className="col-12 col-lg-6 mb-4">
                    <div
                        className="card h-100"
                        data-product-id={product.id}
                        data-testid="productCard"
                    >
                        <div className="row no-gutters">
                            <div className="col-md-4">
                                {(product.img && product.img !== '') ?
                                    (
                                        <img
                                            src={product.img}
                                            className="card-img card-img-left"
                                            alt={product.name}
                                            data-testid="productCardImage"
                                        />
                                    ) : (
                                        <img
                                            src="https://via.placeholder.com/500x900?text=Soon"
                                            className="card-img card-img-left"
                                            alt={product.name}
                                            data-testid="productCardNoImage"
                                        />
                                    )

                                }
                            </div>
                            <div className="col-md-8">
                                <div className="card-body">
                                    <div className="productInfo mb-3">
                                        <h2
                                            className="h3 card-title"
                                            data-testid="productCardTitle"
                                        >
                                            {product.name}
                                        </h2>
                                        <p
                                            className="text-muted small"
                                            data-testid="productCardColour"
                                        >
                                            {product.colour}
                                        </p>
                                        <p
                                            className="price"
                                            data-testid="productCardPrice"
                                        >
                                            {Constant.CURRENCY}{product.price}
                                        </p>
                                    </div>
                                    <div className="qtyContainer mb-3 text-center">
                                        <div className="input-group">
                                            <div className="input-group-prepend">
                                                <button
                                                    onClick={() => this.updateCart('decrease', product.id, product.price)}
                                                    className="btn btn-outline-secondary"
                                                    type="button"
                                                    disabled={disabledState}
                                                    data-testid="productCardDecrease"
                                                >
                                                    <i className="fas fa-minus"></i>
                                                </button>
                                            </div>
                                            <input
                                                type="text"
                                                className="form-control text-center"
                                                placeholder="Qty"
                                                value={qty}
                                                readOnly={true}
                                                data-testid="productCardQtyInput"
                                            />
                                            <div className="input-group-append">
                                                <button
                                                    onClick={() => this.updateCart('increase', product.id, product.price)}
                                                    className="btn btn-outline-secondary"
                                                    type="button"
                                                    data-testid="productCardIncrease"
                                                >
                                                    <i className="fas fa-plus"></i>
                                                </button>
                                            </div>
                                        </div>
                                        <button
                                            onClick={() => this.updateCart('remove', product.id, product.price)}
                                            className="btn btn-link btn-sm"
                                            disabled={disabledState}
                                        >
                                            Remove
                                        </button>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
                }
            </React.Fragment>
        )
    }
}

export default Product;

