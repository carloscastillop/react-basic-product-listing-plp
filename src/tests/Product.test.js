import React from 'react';
import ReactDOM from 'react-dom';
import {render, cleanup} from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

import Product from '../components/Product';

afterEach(cleanup);

/**
 * Basic render
 */
it('Product card renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Product/>, div);
});

/**
 * Render simple product card
 */
it('Product card renders', () => {
    const product = {
        colour: "Black",
        id: 1,
        img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
        name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
        price: 10
    };
    const cart = [{id: 1, price: 10, qty: 10}];
    const {getByTestId} = render(<Product product={product} cart={cart}/>)
    expect(getByTestId('productCardTitle')).toHaveTextContent('Black Sheet Strappy Textured Glitter Bodycon Dress');
    expect(getByTestId('productCardColour')).toHaveTextContent('Black');
    expect(getByTestId('productCardPrice')).toHaveTextContent('£10');
    expect(getByTestId('productCardQtyInput')).toHaveValue('10');

});

/**
 * Render product card no image (A)
 */
it('Product card renders NO image', () => {
    const product = {
        colour: "Black",
        id: 1,
        img: '',
        name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
        price: 10
    };
    const cart = [{id: 1, price: 10, qty: 10}];
    const {getByTestId} = render(<Product product={product} cart={cart}/>)
    expect(getByTestId('productCardTitle')).toHaveTextContent('Black Sheet Strappy Textured Glitter Bodycon Dress');
    expect(getByTestId('productCardColour')).toHaveTextContent('Black');
    expect(getByTestId('productCardPrice')).toHaveTextContent('£10');
    expect(getByTestId('productCardNoImage')).toBeInTheDocument;
});


/**
 * Snapshot (A)
 */
it('matches snapshot A', () => {
    const product = {
        colour: "Black",
        id: 1,
        img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
        name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
        price: 10
    };
    const cart = [{id: 1, price: 10, qty: 10}];
    const tree = renderer.create(<Product product={product} cart={cart}/>).toJSON();
    expect(tree).toMatchSnapshot();
});

/**
 * Functions
 */
describe('<Product /> functions', () => {
    it('qtyFilter', () => {
        const product = {
            colour: "Black",
            id: 1,
            img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
            name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
            price: 10
        };
        const cart = [{id: 1, price: 10, qty: 10}];
        const wrapper = renderer.create(<Product product={product} cart={cart}/>);
        const inst = wrapper.getInstance();
        expect(inst.qtyFilter(product, cart)).toBe(10);
    });

    it('updateCart', () => {
        const {option, id, price} = [{option: 'remove', id: 1, price: 10}];
        const wrapper = renderer.create(<Product/>);
        const inst = wrapper.getInstance();
        //inst.updateCart(option, id, price);
    });
});