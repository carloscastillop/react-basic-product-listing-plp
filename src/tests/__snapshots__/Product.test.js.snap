// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot A 1`] = `
<div
  className="col-12 col-lg-6 mb-4"
>
  <div
    className="card h-100"
    data-product-id={1}
    data-testid="productCard"
  >
    <div
      className="row no-gutters"
    >
      <div
        className="col-md-4"
      >
        <img
          alt="Black Sheet Strappy Textured Glitter Bodycon Dress"
          className="card-img card-img-left"
          data-testid="productCardImage"
          src="http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024"
        />
      </div>
      <div
        className="col-md-8"
      >
        <div
          className="card-body"
        >
          <div
            className="productInfo mb-3"
          >
            <h2
              className="h3 card-title"
              data-testid="productCardTitle"
            >
              Black Sheet Strappy Textured Glitter Bodycon Dress
            </h2>
            <p
              className="text-muted small"
              data-testid="productCardColour"
            >
              Black
            </p>
            <p
              className="price"
              data-testid="productCardPrice"
            >
              Â£
              10
            </p>
          </div>
          <div
            className="qtyContainer mb-3 text-center"
          >
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              >
                <button
                  className="btn btn-outline-secondary"
                  data-testid="productCardDecrease"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <i
                    className="fas fa-minus"
                  />
                </button>
              </div>
              <input
                className="form-control text-center"
                data-testid="productCardQtyInput"
                placeholder="Qty"
                readOnly={true}
                type="text"
                value={10}
              />
              <div
                className="input-group-append"
              >
                <button
                  className="btn btn-outline-secondary"
                  data-testid="productCardIncrease"
                  onClick={[Function]}
                  type="button"
                >
                  <i
                    className="fas fa-plus"
                  />
                </button>
              </div>
            </div>
            <button
              className="btn btn-link btn-sm"
              disabled={false}
              onClick={[Function]}
            >
              Remove
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
